### DISCUSSION TITLE
Add Inline Table

### DISCUSSION THREAD

**Discussion by SwingWatcher** [jheddings/notional#65] [Q&A] [answered]
I'm trying to create an inline table(not database table) using the code below but I've got an error. Can someone tell me how to do it?

```
    table = blocks.Table()
    row = blocks.TableRow["1| test"]
    task += table.append(row)
```

I encountered the error below.
```
NotImplementedError                       Traceback (most recent call last)
Input In [64], in <cell line: 71>()
     73 #task.Timeframe = "90M"
     74 #task.Status = "IP"    
     75 #task.Parameters = "NATR=3.15\nFract=2.5"
   (...)
     83 #task += blocks.Heading1["Welcome!"]
     84 #task += blocks.Paragraph["Welcome to the matrix."]
     85 table = blocks.Table()
---> 86 row = blocks.TableRow["1| test"]
     87 table.append(row)

File ~\AppData\Local\Programs\Python\Python310\lib\site-packages\notional\core.py:82, in ComposableObject.__getitem__(self, params)
     72 """Return the requested class by composing using the given param.
     73 
     74 Types found in `params` will be compared to expected types in the `__compose__`
   (...)
     78 an exception.
     79 """
     81 if not hasattr(self, "__compose__"):
---> 82     raise NotImplementedError(f"{self} does not support object composition")
     84 # XXX if params is empty / None, consider calling the default constructor
     86 compose = self.__compose__

NotImplementedError: <class 'notional.blocks.TableRow'> does not support object composition
```

**Comment by jheddings** [comment]
The TableRow is simply a list of text object.  It does not support the composition (square brackets) format at this time.

To create basic tables currently:

```
table = blocks.Table()
row = blocks.TableRow()

# each call to append adds a cell to the row
row.append("1")
row.append("test")

# all rows must be equal in width
table.append(row)
```

Optionally, you can call `row.append(...)` using a list of `RichTextObject`'s.

**Reply by SwingWatcher** [reply]
I tried adding a table to an existing database record.

```
query = NotionStrategy.query().filter(property="Strategy", rich_text=TextCondition(equals="@NQ-123"))

for task in query.execute():
    table = blocks.Table()
    row = blocks.TableRow()
    row.append("ID")
    row.append("1")
    row.append("2")
    table.append(row)
    task += table
```
But I encountered this error.

```
AttributeError                            Traceback (most recent call last)
Input In [3], in <cell line: 72>()
    103 table = blocks.Table()
    104 row = blocks.TableRow()
--> 105 row.append("ID")
    106 row.append("1")
    107 row.append("2")

File ~\AppData\Local\Programs\Python\Python310\lib\site-packages\notional\blocks.py:592, in TableRow.append(self, text)
    589     self.table_row.cells.append([text])
    591 else:
--> 592     rtf = TextObject.from_value(text)
    593     self.table_row.cells.append([rtf])

AttributeError: type object 'TextObject' has no attribute 'from_value'
```

**Reply by jheddings** [reply]
Yeah...  You are running into the bug I ended up fixed in #66  😳 

I published a new package (v0.4.3) that should fix this for you.

**Reply by SwingWatcher** [reply]
> Yeah... You are running into the bug I ended up fixed in #66 😳
> 
> I published a new package (v0.4.3) that should fix this for you.

Thanks I'll update it.

Do you have an example of a table with multiple columns and rows?

**Comment by jheddings** [comment]
Check out the unit test in `test_tables.py` and see if that gets you headed in the right direction.Note that all rows in a table must have the same number of cells.On Nov 8, 2022, at 19:23, Swing Watcher ***@***.***> wrote:﻿

Yeah... You are running into the bug I ended up fixed in #66 😳
I published a new package (v0.4.3) that should fix this for you.

Thanks I'll update it.
Do you have an example of a table with multiple columns and rows?

**Comment by jheddings** [comment]
Looks like you are creating a new `row1` object, but still appending to the first `row`.  That puts 4 cells in the first instance, but 0 in the `row1` object.

**Reply by SwingWatcher** [reply]
Thanks! It worked.

How can I set a row header and column header? This code didn't work.

```
    table = blocks.Table()
    table.has_row_header = True
    table.has_column_header = True
```

**Comment by jheddings** [comment, marked as answer]
Yeah, it’s a little funky based on the way Notion presents data in the API.  Those parameters are actually part of the nested metadata in the `Table` object:

```
    table = blocks.Table()
    table.table.has_row_header = True
    table.table.has_column_header = True
```

(note the extra `.table` in the property)

**Reply by SwingWatcher** [reply]
Awesome! That's great! Thank you for your help!
